File: c++/combination-sum.cc
class Solution {
public:

  void recursiveBacktrack(const vector<int>& candidates, vector<vector<int>>& combinations, vector<int>& current, int curSum, int index, int target) {
    if (curSum > target || index == candidates.size()) {
      return;
    } else if (curSum == target) {
      combinations.push_back(current);
      return;
      // current.clear();
    }
    current.push_back(candidates[index]);
    recursiveBacktrack(candidates, combinations, current, curSum + candidates[index], index, target);
    current.pop_back();
    recursiveBacktrack(candidates, combinations, current, curSum, index + 1, target);
  }

  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> combinations;
    vector<int> current;
    recursiveBacktrack(candidates, combinations, current, 0, 0, target);

    return combinations;
  }
};

File: python/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        combinations = []

        def recursiveCombinationSum(startIndex: int = 0, currSet: List[int] = [], currSum: int = 0):
            if currSum > target:
                return
            for index in range(startIndex, len(candidates)):
                currSet.append(candidates[index])
                currSum += candidates[index]
                if currSum == target:
                    combinations.append(currSet[:])
                recursiveCombinationSum(index, currSet[:], currSum)
                # recursiveCombinationSum(index + 1, currSet[:], currSum)
                currSum -= candidates[index]
                currSet.pop()
        recursiveCombinationSum()
        return combinations
